services/QuillLMS/app/workers/identify_stripe_invoices_without_subscriptions_worker.rb

```diff
@@ -29,14 +29,35 @@ def perform
   private def relevant_stripe_invoices
     stripe_invoices.filter do |invoice|
       RELEVANT_INVOICE_STATUSES.include?(invoice.status) &&
-        !Subscription.exists?(stripe_invoice_id: invoice.id)
+        !invoice_ids_with_subscriptions.include?(invoice.id) &&
+        !invoice_refunded?(invoice) &&
+        non_zero_amount?(invoice)
     end
   end
 
+  private def invoice_ids_with_subscriptions
+    @invoice_ids_with_subscrirptions ||= Subscription.where(stripe_invoice_id: stripe_invoice_ids).pluck(:stripe_invoice_id)
+  end
+
+  private def invoice_refunded?(invoice)
+    return false unless invoice.charge
+
+    charge = Stripe::Charge.retrieve(invoice.charge)
+    charge.amount == charge.amount_refunded
+  end
+
+  private def non_zero_amount?(invoice)
+    invoice.amount_due > 0
+  end
+
+  private def stripe_invoice_ids
+    stripe_invoices.map(&:id)
+  end
+
   private def stripe_invoices
-    [].tap do |stripe_invoices|
+    @stripe_invoices ||= [].tap do |stripe_invoices|
```

Comment:
For clarity, I would use a different variable name for the tap variable (e.g. `invoices`) than the method name